### PR TITLE
on flow api start recrypt peer configs not encrypted with current key

### PR DESCRIPTION
**Pull request by serprex**

### CODE REVIEW

File: flow/cmd/api.go [hunk 1]
```diff
@@ -168,13 +174,72 @@ func APIMain(ctx context.Context, args *APIServerParams) error {
...
-		if err := gateway.ListenAndServe(); err != nil {
+		if err := gateway.ListenAndServe(); err != nil && err != http.ErrServerClosed {
 			log.Fatalf("failed to serve http: %v", err)
 		}
 	}()
 
-	<-ctx.Done()
+	// somewhat unrelated, need a process which isn't replicated
+	// migrate peer protos to current encryption key
+	go func() {
```
Comment (iskakaushik) [inline flow/cmd/api.go:184]: can we do this before flow worker establishes a connection w/ temporal etc? Can we add info logs around:
1. If we are skipping or current key id is empty.
2. previous and current key id for a peer.
3. Successfully encrypted.

Comment (serprex) [reply flow/cmd/api.go]: discussed aside to leave async operation

File: flow/cmd/api.go [hunk 2]
```diff
@@ -37,6 +39,97 @@ type APIServerParams struct {
...
+		slog.Warn("recrypt failed to find key, skipping", slog.Any("error", err))
+		return
+	}
+
+	tx, err := catalogPool.BeginTx(ctx, pgx.TxOptions{})
+	if err != nil {
+		slog.Warn("recrypt failed to start transaction, skipping", slog.Any("error", err))
+		return
+	}
+	defer func() {
```
[the remark this answers is not in the document]
Comment (serprex) [reply flow/cmd/api.go:69]: that relies on temporal logger

File: flow/cmd/api.go [hunk 3]
```diff
@@ -37,6 +39,97 @@ type APIServerParams struct {
...
+	var oldKeyID string
+	for rows.Next() {
+		if err := rows.Scan(&id, &options, &oldKeyID); err != nil {
+			slog.Warn("recrypt failed to scan, skipping", slog.Any("error", err))
+			continue
+		}
+
+		oldKey, err := keys.Get(oldKeyID)
+		if err != nil {
+			slog.Warn("recrypt failed to find key, skipping", slog.Any("error", err), slog.String("enc_key_id", oldKeyID))
```
[the remark this answers is not in the document]
Comment (serprex) [reply flow/cmd/api.go:92]: it'll happen if keys break or field somehow corrupt

File: flow/cmd/api.go [hunk 4]
```diff
@@ -37,6 +39,97 @@ type APIServerParams struct {
...
+}
+
+func recryptDatabase(ctx context.Context, catalogPool *pgxpool.Pool) {
+	newKeyID := peerdbenv.PeerDBCurrentEncKeyID()
+	keys := peerdbenv.PeerDBEncKeys()
+	if newKeyID == "" {
+		if len(keys) == 0 {
+			slog.Warn("Encryption disabled. This is not recommended.")
+		} else {
+			slog.Warn("Encryption disabled, decrypting any currently encrypted configs. This is not recommended.")
```
[the remark this answers is not in the document]
Comment (serprex) [reply flow/cmd/api.go:54]: could check if env variable is explicitly empty, but for now leaving as is

**Pull request merged**